Remove Buffer and File class friends

derobins opened the pull request:
I don't recall the reason for making the ctors private, requiring friend classes to instantiate them. I'm personally not a fan of breaking encapsulation with friend classes.

derobins committed:
Remove buffer and file friends

I don't recall the reason for making the ctors private, requiring
friend classes to instantiate them. I'm personally not a fan of
breaking encapsulation with friend classes.

jordan-turbofish commented:
As they are now, File and Buffer objects are guaranteed to be registered.  This change would make that a convention, rather than enforced by the compiler.

kurtmcmillan commented:
> As they are now, File and Buffer objects are guaranteed to be registered. This change would make that a convention, rather than enforced by the compiler.

I was making this point when you made your comment. There is one exception, we temporarily create a Buffer for a hipFileRead/hipFileWrite if the buffer is not registered.

derobins committed:
Use PassKey idiom in File/Buffer/Map classes

Prevents unrestricted friendship between the File & Buffer classes
and FileMap & BufferMap, respectively.

Needs some additional documentation since this is an uncommon
pattern.

derobins committed:
Add a helpful note about the Key class

derobins committed:
Add Doxygen for the new ctor params

derobins commented:
Article about this pattern here:

https://arne-mertz.de/2016/10/passkey-idiom/

And here:

https://stackoverflow.com/questions/3217390/clean-c-granular-friend-equivalent-answer-attorney-client-idiom/3218920

derobins committed:
Explicitly mark ctors private

riley-dixon reviewed (commented):
Looks good to me. But I think Kurt/Jordan should give a thumbs up for this to go in.

derobins commented:
I'm going to take a sec to rename Key to PassKey to emphasize that it's not a key like in a KV store or anything.

jordan-turbofish commented:
Do you want to make the same change to stream too?

derobins committed:
Rename Key --> PassKey

derobins committed:
Update the Stream class to work like File/Buffer

derobins committed:
Fix key.h --> passkey.h in comments

derobins committed:
Make PassKeys const
